### PR TITLE
Pipeline implementation of learning algorithm

### LINKED ISSUE #177: Clean up options passing
Can pass around as e.g. an object or a dictionary, with user-supplied values merged into a default set

For cleaning up e.g. the log reg functions

### LINKED ISSUE #232: Finish pipeline version with option to return marginals after first stage
So these can be passed to the LSTM module

@henryre is this almost done?  @SenWu

### PR DESCRIPTION
**Pull request by henryre**
Fixes #232 
Fixes #177

### CODE REVIEW

**Comment by henryre** [comment]
Thanks Travis...

Removing the current learning tests, since they should be rewritten (see #226)

**Comment by henryre** [comment]
^ eh jk I'll just patch the current ones

**Comment by henryre** [comment]
Oy and I need to update the docs. Hang tight on this one

**Comment by henryre** [comment]
Adding @SenWu 's suggestion...
Fixes #237

**Comment by ajratner** [comment]
Cool checking out this PR now

**Comment by henryre** [comment]
^ don't merge quite yet

**Comment by ajratner** [comment]
Okay ping again when ready

**Comment by henryre** [comment]
@ajratner ping good to go!

**Comment by henryre** [comment]
This ca. 1972 A320 that jetblue is flying doesn't have plugs, so may cut out suddenly

**Comment by ajratner** [comment]
Sounds good will look at in a bit

File: README.md [hunk 1]
```diff
@@ -264,15 +264,70 @@ Update coming...
...
+|`learn_weights(**kwargs)` | See [Learning in DDLite](#learnopts) below |
 |`learn_weights_validated(**kwargs)` | Wrapper for `learn_weights()`|
 |`set_use_lfs(use=True)` | Setter for using LF weights in prediction |
 |`get_log_odds(subset=None)` | Get log odds values for all candidates (`subset=None`), an index subset (`subset`), test set (`subset='test'`), or validation set (`subset='validation'`) |
 |`get_predicted_probability(subset=None)` | Get predicted probabilities for candidates |
 |`get_predicted(subset=None)` | Get predicted responses for candidates |
+|`get_log_odds_lf(subset=None)` | Get log odds values for all candidates (`subset=None`), an index subset (`subset`), test set (`subset='test'`), or validation set (`subset='validation'`) using only LFs|
```
Comment (ajratner) [inline README.md]: ?  Not clear what this is?  Is this the weights of the LFs, or the marginals for the candidates, or...?

File: README.md [hunk 2]
```diff
@@ -264,15 +264,70 @@ Update coming...
...
+|`get_predicted_probability_lf(subset=None)` | Get predicted probabilities for candidates using only LFs |
+|`get_predicted_lf(subset=None)` | Get predicted responses for candidates using only LFs|
```
Comment (ajratner) [inline README.md]: "predicted probability" -> "Marginal probability", "responses" -> "labels"..?

File: README.md [hunk 3]
```diff
@@ -264,15 +264,70 @@ Update coming...
...
+
 |`get_classification_accuracy(subset=None)` | Get classification accuracy over `subset` against ground truth |
 |`plot_calibration()` | Show DeepDive calibration plots |
 |`open_mindtagger(num_sample=None, abstain=False, **kwargs)` | Open MindTagger portal for labeling; sample is either the last sample (`num_sample=None`) or a random sample of size `num_sample`; if `abstain=True` show only canidates which haven't been labeled by any LF |
 |`add_mindtagger_tags()` | Import tags from current MindTagger session |
 |`add_to_log(log_id=None, subset='test', show=True)` | Add current learning result to log; precision and recall computed from ground truth in `subset` |
 |`show_log(idx=None)` | Show all learning results (`idx=None`) or learning result `idx` | 
+
+## Learning in DDLite<a name="learnopts"></a>
+DDLite provides a number of options for learning predictive models. The simplest option is to just `DDLiteModel.learn_weights()`, but the results probably won't be so hot using only default values. There are two important decisions to make before setting other parameters.
```
Comment (ajratner) [inline README.md]: Remove "won't be so hot"

File: README.md [hunk 4]
```diff
@@ -264,15 +264,70 @@ Update coming...
...
+
+* **Pipeline or joint learning?** In the pipeline learning approach, the LF accuracies are learned separately from the feature weights. These are the "LF stage" and "feature stage", respectively. In the joint learning approach, the LF accuracies and feature weights are learned at the same time. We suggest using pipeline (`pipeline=True`) since it's more flexible than the joint approach (`pipeline=False`); pipeline is also the default.
```
Comment (ajratner) [inline README.md]: Instead of "learned separately", something indicating sequential, more like "learned first, then..."

O/w can edit this more later

File: README.md [hunk 5]
```diff
@@ -264,15 +264,70 @@ Update coming...
...
+| `w0_mult_lf` | `1` | `float` | Multiplier for initial LF weights (`1` or `3` are typical) | 
+| `alpha_feats` | `alpha` | `float` | Same as `alpha` |
+| `rate` | `0.01` | `float` | Default learning rate for both stages |
+| `rate_lf` | `rate` | `float` | Learning rate for LF stage |
+| `rate_feats` | `rate` | `float` | Learning rate for feature stage |
+| `warm_starts_feats` | `warm_starts` | `bool` | Same as `warm_starts` |
+| `mu_lf` | `1e-7`, non-negative | `float` | Regularization parameter for LF stage |
+|`mu_seq_feats` | `mu` | `float` or array-like, all non-negative | Same as `mu`|
+| `n_mu_feats` | `5` | `int` | Number of `mu` values to fit if `mu` is `None` |
+|`mu_min_ratio_feats` | `1e-6` | `float` | Ratio of smallest to largest `mu` values to fit if `mu` is `None` |
```
Comment (ajratner) [inline README.md]: No need right now but probably a cleaner way to integrate all these duplicate joint/pipeline params...

File: ddlite.py [hunk 6]
```diff
@@ -1132,39 +1135,63 @@ def plot_learning_diagnostics(self, mu_opt, f1_opt):
...
+    start = 0 if self.use_lfs else self.num_lfs()
+    return self._log_odds(self.X, self.w, start, subset)
+    
+  def get_log_odds_lf(self, subset=None):
+    """
+    Get the array of predicted link function values (continuous) given LF weights
+    Return either all candidates, a specified subset, or only validation/test set
+    """
+    return self._log_odds(self.lf_matrix.tocsr(), self._lf_w, 0, subset)
+
```
Comment (ajratner) [inline ddlite.py]: I'm confused about what these mean- `get_log_odds` and `get_log_odds_lf`... are these functions returning the marginal probabilities of the candidates?  If so the naming here is a bit off.  Also I don't know what "link function" means here?

File: test/InferenceTests.py [hunk 7]
```diff
@@ -18,8 +19,8 @@ def logistic_regression_small(self, alpha, mu):
         X, w0, gt = self._problem(n=250, nlf=50, lf_prior=0.4, lf_mean=0.7,
                                   lf_sd=0.25, nf=250, f_prior=0.2, f_mean=0.6,
                                   f_sd=0.25)
-        w = learn_elasticnet_logreg(X, n_iter=500, tol=1e-4, w0=w0,
-                                    sample=False, alpha=alpha, mu_seq=mu, 
+        w = joint_learn_elasticnet_logreg(X=X, pipeline=False, n_iter=500, tol=1e-4,
```
Comment (ajratner) [inline test/InferenceTests.py]: Shouldn't it not be necessary to specify `pipeline=False` here?

**Comment by ajratner** [comment]
@SenWu could you take a look at this commit and see if it gets you what you need?

@henryre Thanks for the help with this first of all.  Overall I think the code here is getting a bit opaque and bloated for such a simple core alg, and some of the terminology is a bit off or confusing; we've discussed a refactor here so let's just clean up a bit for now, then do that?  Happy to discuss next week / intermittently this one!

File: README.md (hunk 1 above)
Comment (henryre) [inline README.md]: marginals for everyone

File: README.md (hunk 2 above)
Comment (henryre) [inline README.md]: Sounds good. wanted to try to keep it consistent for comparison purposes, but i agree

File: README.md (hunk 3 above)
Comment (henryre) [inline README.md]: Haha couldn't finish the sentence so left my place holder... on it

File: ddlite.py (hunk 6 above)
Comment (henryre) [inline ddlite.py]: No, this is all GLM speak. It's X.dot(w), the linear response before mapping to probabilities

File: test/InferenceTests.py (hunk 7 above)
Comment (henryre) [inline test/InferenceTests.py]: `pipeline=True` is default

**Comment by henryre** [comment]
@ajratner Where specifically are you seeing bloat besides the argument passing?

**Comment by ajratner** [comment]
Didn't mean to sound too critical- just a lot of lines of code for LR.  We'll review / I'll try to qualify that statement soon...

**Comment by henryre** [comment]
No, no prob. Let's just get to the bottom of it. We should have a clean version in Python now in case the refactor gets delayed. The args got bloated because I was trying to let people just swap `pipeline=True` and `pipeline=False`, but 1. not a real use case and 2. easier ways to do that. I'll clean that up tomorrow

File: ddlite.py [hunk 8]
```diff
@@ -1063,11 +1064,11 @@ def learn_weights(self, **kwargs):
     result = learn_elasticnet_logreg(F=F[train_cov, :], 
                                      LF=LF[train_cov, :],
                                      w0=w0, unreg=unreg, mu_seq=mu, **kwargs)
-    w_all, self._lf_marginals = result, None    
+    w_all, self._lf_w = result, None    
     if pipeline:
       zw = np.zeros_like(result[0])
       w_all = {k : np.concatenate((zw, v)) for k,v in result[1].iteritems()}
-      self._lf_marginals = result[2]
+      self._lf_w = result[2]
```
Comment (SenWu) [inline ddlite.py]: Currently, **_lf_w** is just a list of marginals and **training()** is the id list of training data. They have the same length, but how to find matching between these two lists.

**Comment by henryre** [comment]
@SenWu try pulling again. `_lf_w` is now the weights associated with the LF accuracies (so `len(_lf_w) == num_lfs()`. The function `get_predicted_probability_lf()` uses these weights and returns the marginal probabilities for each candidate. You can then index the returned array of probabilities however you want.

**Comment by henryre** [comment]
@ajratner ok here's the "three pronged plan" for fixing this up

1. Remove option for joint learning in DDLite. This will remove needless confusion, allow cleaner integration with LSTMs, and remove the parameter passing bloat since we only need checks for calls to the pipeline function
2. Remove the log odds-based functions and instead store a precomputed array of LF marginal probabilities and probabilities output from the final model. This again removes needless confusion, and allows cleaner integration with LSTMs.
3. Disrupt.

**Comment by chrismre** [comment]
I like 3.

**Comment by ajratner** [comment]
Sounds like a good plan to me! I would also add that we should continue
thinking of ways to maintain a separation between the "core alg" (eg
gradient update, etc) and all the parameter tuning bells and whistles code,
just for cleanliness

Otherwise sounds great

As for disruption: i think of this PR as basically deep learning for the
uber for marginal probabilities delivery. So yeah, it's gonna disrupt.

**Comment by ajratner** [comment]
@henryre Let me know when ready to look at; also anything I can help code up?

**Comment by henryre** [comment]
@ajratner Ok take a look at this before I update docs

File: ddlite.py [hunk 9]
```diff
@@ -680,18 +679,18 @@ def __init__(self, candidates, feats=None, gt_dict=None):
...
     self.lf_names = []
-    # Model data    
-    self.X = None
-    self._w_fit = None
-    self.w = None
+    # Model probabilities
+    self.model = None
+    self.probabilities = None
+    self.lf_marginals = None
+    self.lf_weights = None
```
Comment (ajratner) [inline ddlite.py]: Still a little confused here- what's `self.probabilities`?

I think `lf_marginals` should maybe be `training_marginals` or something (otherwise sounds like it's the marginal prob of the LFs)

File: ddlite.py [hunk 10]
```diff
@@ -1017,85 +1019,92 @@ def lf_summary_table(self):
...
+                                              sample=sample,
+                                              n_samples=n_samples,
+                                              verbose=verbose)
+    self.lf_marginals = odds_to_prob(np.ravel(LF.dot(self.lf_weights)))
+    
+  def train_model_lr(self, bias=False, n_mu=5, mu_min_ratio=1e-6, 
+                     plot=True, log=True, **kwargs):
+    """ Learn second stage of pipeline with logistic regression """
+    self.model = "Logistic regression"
+    # Warn if test size is small
```
Comment (ajratner) [inline ddlite.py]: 👍 💯

File: ddlite.py [hunk 11]
```diff
@@ -1157,14 +1158,28 @@ def get_predicted_probability(self, subset=None):
...
     """
-    return np.sign(self.get_log_odds(subset))
+    return np.sign(self.get_predicted_probability(subset) - 0.5)
+    
+  def get_lf_predicted(self, subset=None):
+    """
+    Get the array of predicted (boolean) variables given LF weights
+    Return either all variables, a specified subset, or only validation/test set
+    """
+    return np.sign(self.get_lf_predicted_probability(subset) - 0.5)
```
Comment (ajratner) [inline ddlite.py]: Still think these fn names are misleading, again maybe something like `get_training_marginals` and `get_training_predicted`?

File: ddlite.py (hunk 9 above)
Comment (henryre) [inline ddlite.py]: `probabilities[i]` = P(y_i | \Lambda, x_i) where x_i are the features

The training name is ambiguous to me, I'm ok going with that for now

The weights are not accuracies. They're the preimage of the accuracies under the logit function

**Comment by ajratner** [comment]
@henryre Only minor comments, otherwise as long as we can confirm that the tests run LGTM!

Also, I'm hoping to move documentation to just (1) single (CDR) tutorial, and (ii) auto-generated docs.  So I would just beef up the doc strings (see [this handy, down-to-earth little reference](https://www.python.org/dev/peps/pep-0257/)!), then use `pdoc` as in:
https://github.com/HazyResearch/ddlite/blob/candidate-extractor-2/generate_docs.sh

File: ddlite.py (hunk 9 above)
Comment (ajratner) [inline ddlite.py]: Ok I agree; maybe then stick w what you had just be consistent e.g. `lf_marginals` and `marginals`?  I dunno, I guess whatever is fine

Sure I was just mirroring your title below, that's fine, leave as `lf_weights`

**Comment by henryre** [comment]
I'm just going to merge this in without @ajratner's docstring fanciness to get the code in and stop blocking. Will start on new docs today

**Comment by henryre** [comment]
PEP 8

**Comment by henryre** [comment]
:( wrong platform

**Comment by henryre** [comment]
All checks passed on Henry CI!

**Comment by ajratner** [comment]
PEP-257 sheesh. Sounds good :)

**Comment by ajratner** [comment]
Do add in any post-merge though!
